extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTools.java

```diff
@@ -1639,11 +1853,20 @@ public static Table readSingleFileTable(
     private static Table readSingleFileTable(
             @NotNull final URI parquetFileURI,
             @NotNull final ParquetInstructions readInstructions) {
-        final ParquetSingleFileLayout keyFinder = new ParquetSingleFileLayout(parquetFileURI, readInstructions);
-        final KnownLocationKeyFinder<ParquetTableLocationKey> inferenceKeys = toKnownKeys(keyFinder);
+        verifyFileLayout(readInstructions, ParquetFileLayout.SINGLE_FILE);
+        if (readInstructions.getTableDefinition().isPresent()) {
+            return readSingleFileTable(new ParquetTableLocationKey(parquetFileURI, 0, null, readInstructions),
+                    readInstructions);
+        }
+        // Infer the table definition
+        final TableLocationKeyFinder<ParquetTableLocationKey> singleFileLayout =
+                new ParquetSingleFileLayout(parquetFileURI, readInstructions);
+        final KnownLocationKeyFinder<ParquetTableLocationKey> inferenceKeys = toKnownKeys(singleFileLayout);
         final Pair<TableDefinition, ParquetInstructions> inference = infer(inferenceKeys, readInstructions);
-        return readSingleFileTable(inferenceKeys.getFirstKey().orElseThrow(), inference.getSecond(),
-                inference.getFirst());
+        final TableDefinition inferredTableDefinition = inference.getFirst();
+        final ParquetInstructions inferredInstructions = inference.getSecond();
+        return readSingleFileTable(inferenceKeys.getFirstKey().orElseThrow(),
+                ensureTableDefinition(inferredInstructions, inferredTableDefinition));
```

Comment:
I think we could share more code if the method formerly known as `readPartitionedTable` has a switch on layout used to construct a key finder. Or maybe not worth it, since it also needs to consider layout for the result table type to construct.